weave-js/src/components/PagePanelComponents/Home/Browse3/pages/CompareEvaluationsPage/sections/ExampleCompareSection/exampleCompareSectionUtil.ts

```diff
@@ -323,9 +309,183 @@ export const useFilteredAggregateRows = (state: EvaluationComparisonState) => {
+async function makePartialTableReq(
+  evaluations: UseExampleCompareDataParams[0]['summary']['evaluations'],
+  filteredRows: UseExampleCompareDataParams[1],
+  targetIndex: UseExampleCompareDataParams[2],
+  getTraceServerClient: () => TraceServerClient
+) {
+  const targetRow = filteredRows[targetIndex];
+  if (targetRow == null) {
+    return null;
+  }
+  const datasetRef = Object.values(evaluations)[0].datasetRef as string;
+
+  const datasetObjRes = await getTraceServerClient().readBatch({
+    refs: [datasetRef],
+  });
+  if (!datasetObjRes.vals[0]) {
+    console.error('Dataset not found');
+    return null;
+  }
+
+  const rowsRef = datasetObjRes.vals[0].rows;
+  const parsedRowsRef = parseRef(rowsRef) as WeaveObjectRef;
+
+  return {
+    project_id: projectIdFromParts({
+      entity: parsedRowsRef.entityName,
+      project: parsedRowsRef.projectName,
+    }),
+    digest: parsedRowsRef.artifactVersion,
+  };
+}
+
+async function loadRowDataIntoCache(
+  rowDigests: string[],
+  cachedRowData: MutableRefObject<Record<string, any>>,
+  cachedPartialTableRequest: MutableRefObject<{
+    project_id: string;
+    digest: string;
+  } | null>,
+  getTraceServerClient: () => TraceServerClient
+) {
+  const rowsRes = await getTraceServerClient().tableQuery({
+    ...cachedPartialTableRequest.current!,
+    filter: {
+      row_digests: rowDigests,
+    },
+  });
+  for (const row of rowsRes.rows) {
+    cachedRowData.current[row.digest] = row.val;
+  }
+}
+
+type UseExampleCompareDataParams = Parameters<typeof useExampleCompareData>;
+
+export function useExampleCompareData(
+  state: EvaluationComparisonState,
+  filteredRows: Array<{
+    inputDigest: string;
+  }>,
+  targetIndex: number
+) {
+  const getTraceServerClient = useGetTraceServerClientContext();
+
+  // cache the row data for the current target row and adjacent rows,
+  // this is to allow for fast re-renders during pagination
+  const cachedRowData = useRef<Record<string, any>>({});
+  const cachedPartialTableRequest = useRef<{
+    project_id: string;
+    digest: string;
+  } | null>(null);
+
+  // This is to provide a way to manually control the re-render of the target row
+  const [cacheVersion, setCacheVersion] = useState<number>(0);
+  const increaseCacheVersion = useCallback(() => {
+    setCacheVersion(prev => prev + 1);
+  }, []);
+
+  const [loading, setLoading] = useState<boolean>(false);
+
+  const targetRowValue = useMemo(() => {
+    if (isEmpty(filteredRows)) {
+      return undefined;
+    }
+    const digest = filteredRows[targetIndex].inputDigest;
+    return flattenObjectPreservingWeaveTypes(cachedRowData.current[digest]);
+    // Including `cacheVersion` in the dependency array ensures the memo recalculates
+    // when it changes, even though it's not directly used in the calculation.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [cacheVersion, filteredRows, targetIndex]);
```

Comment:
I am not the biggest fan of this pattern (in particular, explicitly depending on `cacheVersion` that is not used in the hook). But I would agree it is hard to get around this one. Approving, but do think there might be a cleaner way.